### PR TITLE
Parse SBAT variable and perform basic verification

### PR DESCRIPTION
**Pull request by aburmash**
Per @vathpela suggestion, we will be flexible in what data we expect while parsing the variable.
 Three fields are mandatory:
 component_generation, component_name_size, component_name
 However we also support adding comments and additional information to be added after component name, with
 ',' as a separator. Those information will be ignored and not used for verification purposes.
 So:
 1,5,grub\n
 1,5,grub,wow,this,is,my,comment\n
 will provide exactly same set of data for verification.

Reference: https://github.com/rhboot/shim/blob/main/SBAT.md

Signed-off-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Alex Burmashev <alexander.burmashev@oracle.com>

### CODE REVIEW

File: include/sbat.h [hunk 1]
```diff
@@ -8,6 +8,16 @@
```
Comment (martinezjavier) [inline include/sbat.h:8]: I wish Github had a way to comment on the commit messages, so I'll comment here.

shim me more or less follow what's described in https://www.kernel.org/doc/html/v4.17/process/submitting-patches.html#the-canonical-patch-format, please flollow that. Specially line wrapping at ~75 columns.

Also I think it's better if you use Peter Jones instead of @vathpela in the commit message.

File: include/sbat.h [hunk 2]
```diff
@@ -22,8 +32,12 @@ struct sbat {
 	struct sbat_entry **entries;
 };
 
+int parse_sbat_var(list_t *entries);
+
+EFI_STATUS verify_sbat(struct sbat *sbat, list_t *entries);
+
 EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
-		      struct sbat *sbat);
+                      struct sbat *sbat);
```
Comment (martinezjavier) [inline include/sbat.h:40]: This seems to be an unrelated change.

File: sbat.c [hunk 3]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+			FreePool((CHAR8 *)entry->component_generation);
+		if (entry->component_name_size)
+			FreePool((CHAR8 *)entry->component_name_size);
+		if (entry->component_name)
+			FreePool((CHAR8 *)entry->component_name);
+		FreePool(entry);
+	}
+}
+
+int
```
Comment (martinezjavier) [inline sbat.c:210]: I would use `EFI_STATUS` instead here. That way you could propagate the errors returned by EFI services calls and also would be consistent with other functions in this file.

File: pe.c [hunk 4]
```diff
@@ -1074,18 +1075,23 @@ handle_image (void *data, unsigned int datasize,
 				       entry->vendor_version,
 				       entry->vendor_url);
 			}
+			res = parse_sbat_var(&sbat_var_entries);
+			if (res < 0)
+				console_print(L"could not parse SBAT variable");
```
Comment (martinezjavier) [inline pe.c:1080]: shouldn't this be an error instead of just an info log? And the error propagated ? This is another reason to make this function return an `EFI_STATUS`.

File: pe.c [hunk 5]
```diff
@@ -1074,18 +1075,23 @@ handle_image (void *data, unsigned int datasize,
...
 		} else {
 			perror(L"SBAT data not found\n");
 			return EFI_UNSUPPORTED;
 		}
-
```
Comment (martinezjavier) [inline pe.c:1081]: Please drop this unrelated change.

File: pe.c [hunk 6]
```diff
@@ -1074,18 +1075,23 @@ handle_image (void *data, unsigned int datasize,
...
 		efi_status = verify_buffer(data, datasize,
 					   &context, sha256hash, sha1hash);
-
+		if (sbat.entries)
```
Comment (martinezjavier) [inline pe.c:1087]: We should check for this only one.

File: pe.c [hunk 7]
```diff
@@ -1074,18 +1075,23 @@ handle_image (void *data, unsigned int datasize,
...
+			efi_status = verify_sbat(&sbat, &sbat_var_entries);
+			/* Add EFI_SECURITY_VIOLATION if variable is set,
+			and binary has no SBAT section ?*/
 		if (sbat.entries)
```
Comment (martinezjavier) [inline pe.c:1091]: Drop this and make it all part of the same block if `sbat.entries` is set. This condition should always be true anyways.

File: pe.c [hunk 8]
```diff
@@ -1074,18 +1075,23 @@ handle_image (void *data, unsigned int datasize,
...
 		} else {
 			perror(L"SBAT data not found\n");
 			return EFI_UNSUPPORTED;
 		}
-
 		efi_status = verify_buffer(data, datasize,
 					   &context, sha256hash, sha1hash);
-
+		if (sbat.entries)
+			efi_status = verify_sbat(&sbat, &sbat_var_entries);
```
Comment (martinezjavier) [inline pe.c:1088]: you have to propagate the error here.

File: pe.c [hunk 9]
```diff
@@ -1074,18 +1075,23 @@ handle_image (void *data, unsigned int datasize,
...
+			/* Add EFI_SECURITY_VIOLATION if variable is set,
+			and binary has no SBAT section ?*/
 		if (sbat.entries)
 			for (i = 0; i < sbat.size; i++)
 				FreePool(sbat.entries[i]);
-
+
```
Comment (martinezjavier) [inline pe.c:1094]: Please drop all these unrelated changes.

File: sbat.c [hunk 10]
```diff
@@ -4,27 +4,28 @@
  */
 
 #include "sbat.h"
+#include <string.h>
 
 CHAR8 *
-get_sbat_field(CHAR8 *current, CHAR8 *end, const CHAR8 ** field, char delim)
+get_sbat_field(CHAR8 *current, CHAR8 *end, const CHAR8 **field, char delim)
```
Comment (martinezjavier) [inline sbat.c:10]: I see what happened. You also squashed @vathpela's patch that was cleaning up the formatting issues. That should be left as a separate patch and I think that can already be merged.

File: sbat.c [hunk 11]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
 }
 
+EFI_STATUS
+verify_sbat(struct sbat *sbat, list_t *sbat_entries)
+{
+	unsigned int i;
+	list_t *pos = NULL, *tmp = NULL;
+
+	for (i = 0; i < sbat->size; i++) {
+		struct sbat_entry *entry = entry = sbat->entries[i];
```
Comment (martinezjavier) [inline sbat.c:133]: Maybe add a `verify_sbat_entry()` function or something to make this function smaller and easier to read?

File: sbat.c [hunk 12]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+		for_each_sbat_var_safe (pos, tmp, sbat_entries) {
```
Comment (martinezjavier) [inline sbat.c:134]: The `_safe` variant is only needed when you will modify the list / remove elements, but just to iterate you can use the `for_each_sbat_var()` variant.

File: sbat.c [hunk 13]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+			if (strcmp(entry->component_name,
+			           sbat_var_entry->component_name) == 0) {
+				dprint(L"component %a has a matching SBAT variable entry, verifying\n", entry->component_name);
+				/* atoi returns zero for failed conversion, so essentially
+				   badly parsed component_generation will be treated as zero 
+				*/
+				UINT16 sbat_gen = atoi(entry->component_generation);
+				UINT16 sbat_var_gen = atoi(sbat_var_entry->component_generation);
+				if ( sbat_gen < sbat_var_gen ) {
+					dprint(L"component's %a generation: %d. Conflicts with SBAT variable generation %d\n",
```
Comment (martinezjavier) [inline sbat.c:146]: not sure I understand this debug message. What exactly conflicts means in this context ?

Maybe something like "component %a generation %d has been revoked by SBAT variable" or something like that?

File: sbat.c [hunk 14]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+{
+	unsigned char bom[] = { 0xEF, 0xBB, 0xBF };
+	return !CompareMem(buf, bom, MIN(sizeof(bom), bufsize));
+}
+
+static struct sbat_var *
+new_entry(const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
```
Comment (martinezjavier) [inline sbat.c:169]: You need to check for allocation failures.

File: sbat.c [hunk 15]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+
+	INIT_LIST_HEAD(&new_entry->list); 
+	new_entry->component_generation = comp_gen;
+	new_entry->component_name_size = comp_name_size;
+	new_entry->component_name = comp_name;
+	return new_entry;
+}
+
+static int
```
Comment (martinezjavier) [inline sbat.c:178]: Maybe returning `EFI_STATUS` here as well ?

File: sbat.c [hunk 16]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+add_entry(list_t *list, const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new;
+	new = new_entry(comp_gen, comp_name_size, comp_name);
+	if (!new)
+		return -1;
```
Comment (martinezjavier) [inline sbat.c:185]: `return EFI_OUT_OF_RESOURCES` ?

File: sbat.c [hunk 17]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+}
+
+static void
+clean_up_vars(list_t *entries)
+{
+	list_t *pos = NULL, *tmp = NULL;
+	for_each_sbat_var_safe(pos, tmp, entries) {
+		struct sbat_var *entry;
+
+		entry = list_entry(pos, struct sbat_var, list);
```
Comment (martinezjavier) [inline sbat.c:198]: I think `struct sbat_var *entry = list_entry(pos, struct sbat_var, list);` is easier to read.

File: sbat.c [hunk 18]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+		list_del(&entry->list);
+		if (entry->component_generation)
+			FreePool((CHAR8 *)entry->component_generation);
```
Comment (martinezjavier) [inline sbat.c:201]: `FreePool()` takes `void *` as a parameter, I don't think there's a need to cast there.

File: sbat.c [hunk 19]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+{
+	UINT8 *data = 0;
+	UINTN datasize, i;
+	EFI_STATUS efi_status;
+	int rc;
+
+	INIT_LIST_HEAD(entries); 
+	efi_status = get_variable(L"SBAT", &data, &datasize, SHIM_LOCK_GUID);
+	if (EFI_ERROR(efi_status)) {
+		return -1;
```
Comment (martinezjavier) [inline sbat.c:221]: `return efi_status`

File: sbat.c [hunk 20]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+	dprint(L"SBAT variable data:\n");
+
+	rc = 0;
+	char delim;
+	while (start) {
+		const CHAR8 *fields[3] = { NULL, };
+		for (i = 0; i < 4; i++) {
+			const CHAR8 *tmp;
+			/* on fourth iteration we check if we had extra stuff on line while parsing last meaningful extraction of data
+			   if delimeter on 3rd iteration was ',', this means, we have comments after component_name, so let's drop them
```
Comment (martinezjavier) [inline sbat.c:241]: Please keep lines in less than 80 chars.

File: sbat.c [hunk 21]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+			*/
+			if (i == 3 && start[0] != '\0') {
+				if (delim == ',') {
+					start = get_sbat_field(start, end, &tmp, '\n');
```
Comment (martinezjavier) [inline sbat.c:245]: Please add a comment before this that mentions this is the optional comment field that's being parsed.

File: sbat.c [hunk 22]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+					break;
+				}
+				else
```
Comment (martinezjavier) [inline sbat.c:248]: Why do you need the else clause if you are breaking in both cases ?

Comment (martinezjavier) [reply sbat.c:248]: Also, if we are expecting 3 or 4 fields anyways I'm not sure that implementing it as a for loop is more readable instead of just trying to parse each field individually.

File: sbat.c [hunk 23]
```diff
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+			}
+			delim = ',';
+			/* we do not want to jump to next line and grab stuff from that
+			*/
+			if ( (strchrnula(start,'\n') - start + 1) <= (strchrnula(start,',') - start + 1)) {
+				delim = '\n';
+			}
+			if (!start) {
+				rc = -1;
+				break;
```
Comment (martinezjavier) [inline sbat.c:259]: I think is confusing which break is for the for loop and which one is for the while loop, maybe better to add an `err` or `end` label before the `if (rc < 0)` check and a `goto` statement for the error paths ?

File: sbat.c (hunk 3 above)
Comment (vathpela) [reply sbat.c:210]: Yeah, that's probably true - my bad for suggesting int instead of EFI_STATUS.

File: sbat.c (hunk 11 above)
Comment (aburmash) [reply sbat.c:155]: added at 158,  or did you mean to wrap even more code into that verify_single_entry ?

File: sbat.c [hunk 24]
```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+verify_single_entry(struct sbat_entry *entry, struct sbat_var *sbat_var_entry)
+{
+        UINT16 sbat_gen, sbat_var_gen;
+
+        if (strcmp(entry->component_name, sbat_var_entry->component_name) == 0) {
+                dprint(L"component %a has a matching SBAT variable entry, verifying\n",
+                       entry->component_name);
+                /* atoi returns zero for failed conversion, so essentially
+                badly parsed component_generation will be treated as zero
+                */
```
Comment (martinezjavier) [inline sbat.c:134]: Right, I remember now reading a comment from @vathpela about this. I wonder if we should add a `strtoul()` to shim or use the one provided by Cryptlib ? The other option is to always start at 1 to make sure that we could discriminate between an error and a 0 component generation number.

@jsetje what do you think ?

File: sbat.c (hunk 11 above)
Comment (martinezjavier) [reply sbat.c:155]: That's OK, I think makes the function easier to read.

File: sbat.c [hunk 25]
```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+
+EFI_STATUS
+verify_sbat(struct sbat *sbat, list_t *sbat_entries)
+{
+	unsigned int i;
+	list_t *pos = NULL;
+	EFI_STATUS efi_status;
+	
+	for (i = 0; i < sbat->size; i++) {
+		struct sbat_entry *entry = entry = sbat->entries[i];
```
Comment (martinezjavier) [inline sbat.c:155]: why `struct sbat_entry *entry = entry = sbat->entries[i];` and not just `struct sbat_entry *entry =  sbat->entries[i];` ? 

Probably makes more sense to define the variable outside of the for loop though.

File: sbat.c [hunk 26]
```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+		for_each_sbat_var(pos, sbat_entries) {
+			struct sbat_var *sbat_var_entry = list_entry(pos, struct sbat_var, list);
```
Comment (martinezjavier) [inline sbat.c:157]: Same, let's define these at the beginning of the `verify_sbat()` function.

File: sbat.c [hunk 27]
```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+	dprint(L"all entries from SBAT section verified\n");
+	
+	return efi_status;
+}
+
+static BOOLEAN
+is_utf8_bom(CHAR8 *buf, size_t bufsize)
+{
+	unsigned char bom[] = { 0xEF, 0xBB, 0xBF };
+	return !CompareMem(buf, bom, MIN(sizeof(bom), bufsize));
```
Comment (martinezjavier) [inline sbat.c:172]: I think @vathpela wanted this to be `return CompareMem(buf, bom, MIN(sizeof(bom), bufsize)) != 0; ?

File: sbat.c (hunk 14 above)
Comment (martinezjavier) [reply sbat.c:179]: You missed this comment.

File: sbat.c (hunk 18 above)
Comment (martinezjavier) [reply sbat.c:209]: This one as well, there's no need to cast when calling `FreePool()`.

File: sbat.c [hunk 28]
```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+EFI_STATUS
+parse_sbat_var(list_t *entries)
+{
+	UINT8 *data = 0;
+	UINTN datasize, i;
+	EFI_STATUS efi_status;
+
+	INIT_LIST_HEAD(entries); 
+	efi_status = get_variable(L"SBAT", &data, &datasize, SHIM_LOCK_GUID);
+	if (EFI_ERROR(efi_status)) {
```
Comment (martinezjavier) [inline sbat.c:227]: `if (EFI_ERROR(efi_status) && efi_status != EFI_NOT_FOUND) {`

Comment (martinezjavier) [reply sbat.c:227]: for `EFI_NOT_FOUND` you should just print a message and return `EFI_SUCCESS` I believe, since as you mentioned there may not be a variable at the beginning.

File: sbat.c [hunk 29]
```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
...
+		}
+		dprint(L"component %a with generation %a\n", fields[2], fields[0]);
+		efi_status = add_entry(entries, fields[0], fields[1], fields[2]);
+		if (EFI_ERROR(efi_status))
+			goto error;
+	}
+	FreePool(data);
+	return EFI_SUCCESS;
+error:
+	dprint(L"failed to parse SBAT variable\n");
```
Comment (martinezjavier) [inline sbat.c:283]: Since this is an error condition probably we don't want a debug message and instead a `perror()`.

File: include/str.h [hunk 30]
```diff
@@ -54,7 +54,7 @@ strndupa(const CHAR8 * const src, const UINTN srcmax)
 		return news;
 
 	len = strnlena(src, srcmax);
-	news = AllocateZeroPool(len);
+	news = AllocateZeroPool(len+1);
```
Comment (martinezjavier) [inline include/str.h:57]: `len + 1`. Also this should be a separate patch if is a fix for `strndupa()`.

File: pe.c [hunk 31]
```diff
@@ -1078,13 +1080,24 @@ handle_image (void *data, unsigned int datasize,
...
+		
+		efi_status = parse_sbat_var(&sbat_var_entries);
+		/* we get EFI_SUCCESS only if variable is fully parsed,
+		   and EFI_NOT FOUND means do not perform verificaiton
+		   continue with EFI_SUCCESS
+		*/
+		if (efi_status == EFI_SUCCESS)
+			efi_status = verify_sbat(&sbat, &sbat_var_entries);
+		else if (efi_status == EFI_NOT_FOUND)
+			efi_status = EFI_SUCCESS;
```
Comment (martinezjavier) [inline pe.c:1092]: Don't you need to return an error if the `efi_status` isn't `EFI_SUCCESS` or `EFI_NOT_FOUND` ?

File: pe.c [hunk 32]
```diff
@@ -1078,13 +1080,24 @@ handle_image (void *data, unsigned int datasize,
...
+		
 		efi_status = verify_buffer(data, datasize,
 					   &context, sha256hash, sha1hash);
 
-		if (sbat.entries)
+		if (sbat.entries) {
 			for (i = 0; i < sbat.size; i++)
 				FreePool(sbat.entries[i]);
+		}
```
Comment (martinezjavier) [inline pe.c:1100]: You don't need to add these braces. Although probably makes sense to avoid mistakes later.
